### PR TITLE
Improve torch.ops typing

### PR DESCRIPTION
**Pull request by benjaminglass1**
Fixes longstanding issue where direct references to aten operations are seen as untyped by type checkers. This is accomplished by setting attributes on several classes more consistently, so that `__getattr__` can return a single type in all other cases.

Decisions made along the way:

1. `torch.ops.higher_order` is now implemented by a single-purpose class. This was effectively true before, but the class implementing it attempted to be generalized unnecessarily. Fixing this simplified typing for the `_Ops` class.
2. `__getattr__` is only called when all other lookup methods have failed, so several constant special-cases in the function could be implemented as class variables.

The remainder of this PR is fixing up all the bugs exposed by the updated typing, as well as all the nitpicky typing issues.

Test plan: CI

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -502,7 +502,7 @@ def fmax(self: torch.Tensor, other: torch.Tensor) -> torch.Tensor:
     return torch.where(torch.isnan(other) | (other < self), self, other)
 
 
-@register_decomposition(aten.amax)
+@register_decomposition([aten.amax])
```
Comment (benjaminglass1) [inline torch/_inductor/decomposition.py:505]: This was one of the places where the new typing paid off. `register_decomposition` expects a list of `OperatorBase`/`OpOverloadPacket`. The previous typing of `_OpNamespace` returned `Any` from `__getattr__`, so the implicit call in `aten.amax` couldn't be determined to not be iterable. Once that `__getattr__` was explicitly typed to return `OpOverloadPacket`, typing failed here.

File: torch/_ops.py [hunk 2]
```diff
@@ -1098,14 +1103,22 @@ def _has_script_object_arg(schema: torch.FunctionSchema) -> bool:
 # OpOverloadPacket class contains pointer to a base unresolved operator that doesn't correspond to a specific operator
 # You can obtain an OpOverload object through attribute query.
 class OpOverloadPacket:
-    def __init__(self, qualified_op_name, op_name, op, overload_names):
+    __file__ = "torch.ops"
```
Comment (Skylion007) [inline torch/_ops.py:1106]: This technically a class variable FYI, best to annotate it as such with typing.ClassVar for typing linters.

Comment (benjaminglass1) [reply torch/_ops.py:1106]: @Skylion007 I attempted to do that, but `mypy` actually barfed on it, saying that the base class (`object`, I assume?) has it typed as a `Literal`. That seemed like a weird type to explicitly apply, so I left it alone.

Comment (amjames) [reply torch/_ops.py:1106]: Odd, is it possible that might have been from the `_OpNamespace` type below. It also has a `__file__` but inherits from `typing.ModuleType` which does lead to an error with `mypy` because it is overriding an instance variable as a class var. 
image

I don't see how `__file__` could be inheriting a conflicting annotation from the base type here.

File: torch/_ops.py [hunk 3]
```diff
@@ -1353,17 +1361,19 @@ def _refresh_packet(packet):
     packet._overload_names = overload_names
 
 
-class _PyOpNamespace(_OpNamespace):
```
Comment (amjames) [inline torch/_ops.py:1356]: What is the purpose of this refactor? Is `_PyOpNamespace` single use, so this is more expressive or was there an issue with typing that this helps resolve?

File: torch/fx/node.py [hunk 4]
```diff
@@ -77,15 +77,15 @@
 # Dynamo is unable to trace global set[Callable].__contains__.
 # See https://github.com/pytorch/pytorch/issues/145761. Since we only have
 # a handful of ops so switch to list of callables.
-_side_effectful_need_to_be_preserved_pre_dispatch: list[Callable] = [
+_side_effectful_need_to_be_preserved_pre_dispatch: set[Callable] = {
```
[the remark this answers is not in the document]
Comment (benjaminglass1) [reply torch/fx/node.py]: @amjames Yeah, that caused a lot of pipeline issues (naturally). Already reverted!

File: torch/_ops.py (hunk 3 above)
Comment (benjaminglass1) [reply torch/_ops.py:1356]: OK, there are several things in play here:

1. The base `_OpNamespace` class needs to annotate `__getattr__` with return type `OpOverloadPacket`, for maximum accuracy. This class, OTOH, returns `HigherOrderOperator`, which is an incompatible overload.  To avoid returning an unnecessary `Union` type, and since all but one method in the base was overridden anyway, I figured it was best to break out the class into something standalone.
2. This class was single use, so after breaking it out into its own class, I was able to remove a lot of unneeded complexity by just specializing it.
3. Once the class was appropriately specialized, it became obvious that it could be used once as an explicit member of `_Ops`, which simplified the typing for `_Ops.__getattr__`.

File: torch/_ops.py (hunk 2 above)
Comment (benjaminglass1) [reply torch/_ops.py:1106]: Oh, good point, I may have gotten those confused while typing. I'll give it a shot!

File: torch/_inductor/fx_passes/reinplace.py [hunk 5]
```diff
@@ -176,7 +176,11 @@ def _decompose_scatter_mutating(
 
 def scatter_always_uses_mutation(node: torch.fx.Node) -> bool:
     _, _, view_ops = node.args
-    return any(view.target in _ALWAYS_MUTATING_SCATTER_OPS for view in view_ops)  # type: ignore[union-attr]
+    return any(
+        t in _ALWAYS_MUTATING_SCATTER_OPS
+        for view in view_ops  # type: ignore[union-attr]
+        if isinstance(t := view.target, torch._ops.OpOverload)  # type: ignore[union-attr]
```
Comment (rec) [inline torch/_inductor/fx_passes/reinplace.py:182]: Quibble: adding the intermediate `t` variable doesn't make the code any clearer.

----

FYI: I don't think this works well here but I've been using `TypeGuard` for similar errors in loops, [like this](https://github.com/pytorch/pytorch/blob/e59188cd5c65b35a6437c90fa062aeb25278d268/torch/_inductor/ir.py#L4922-L4925)

File: torch/_ops.py [hunk 6]
```diff
@@ -735,7 +745,14 @@ def get_cached_ops():
 # Each OpOverload object contains pointer to a specific operator overload, a pointer to the parent `OpOverloadPacket` object.
 # You can obtain an OpOverload object through attribute query on OpOverloadPacket.
 class OpOverload(OperatorBase):
-    def __init__(self, overloadpacket, op, op_dk, schema, tags):
+    def __init__(
+        self,
+        overloadpacket: "OpOverloadPacket",
+        op: Callable,
```
Comment (rec) [inline torch/_ops.py:751]: These callables are a bit naked! :-D 

A little surprised this works, I seem to remember seeing errors for naked `Callable` in pytorch, perhaps my memory is mistaken?

Are you expecting the same type in `op` and `op_dk`? If not, the _theoretical_ best way to do this is here:

```
_P = ParamSpec("_P")
_Q = ParamSpec("_Q")

_T = TypeVar("_T")
_U = TypeVar("_U")

# ...

    def __init__(
        op: Callable[_P, _T],
        op_dk: Callable[_Q, _U],
```

What this will do is deduce the parameter and return types when this constructor is actually called, and then potentially detect an issue if these functions are miscalled in the body of the constructor.

In this case, it really has little value because the constructor does nothing, but I do that by habit anyway.

File: torch/_ops.py [hunk 7]
```diff
@@ -1098,14 +1112,22 @@ def _has_script_object_arg(schema: torch.FunctionSchema) -> bool:
...
 # You can obtain an OpOverload object through attribute query.
 class OpOverloadPacket:
-    def __init__(self, qualified_op_name, op_name, op, overload_names):
+    __file__: ClassVar[str] = "torch.ops"
+
+    def __init__(
+        self,
+        qualified_op_name: str,
+        op_name: str,
+        op: Callable,
```
Comment (rec) [inline torch/_ops.py:1121]: See quibble about naked callables above.

File: torch/ao/quantization/pt2e/utils.py [hunk 8]
```diff
@@ -86,7 +86,8 @@ def _find_q_dq_node_for_user(
     q_node = None
     if (
         dq_node.args[0].op == "call_function"  # type: ignore[union-attr]
-        and dq_node.args[0].target in _QUANTIZE_OPS  # type: ignore[union-attr]
+        and isinstance(target := dq_node.args[0].target, torch._ops.OpOverload)  # type: ignore[union-attr]
```
Comment (rec) [inline torch/ao/quantization/pt2e/utils.py:89]: In this one case, you could get rid of the `# type: ignore[union-attr]` with an additional check in the if statement for the type of `dq_node.args[0]`.

File: torch/fx/node.py [hunk 9]
```diff
@@ -85,7 +85,7 @@
 
 # TODO: Either refactor this into 2 functions 1 dce for functional graphs and 1 dce for all graphs,
 # or add logic to correctly mark all inplace ops as side effectful.
-_side_effectful_functions: set[Callable] = {
+_side_effectful_functions: set[Callable] = {  # type: ignore[assignment]
     torch._assert,
```
Comment (rec) [inline torch/fx/node.py:89]: Wait, why did this new ignore appear? A bit dodgy, all of these entries appear to be callable?

File: torch/fx/passes/reinplace.py [hunk 10]
```diff
@@ -252,6 +253,7 @@ def matching_view_metadata(a, b):
         assert isinstance(base.meta["fake_result"], FakeTensor)
         assert isinstance(mutated_view, Node)
         assert isinstance(mutated_view.meta["fake_result"], FakeTensor)
+        assert not isinstance(n.target, str)
```
Comment (rec) [inline torch/fx/passes/reinplace.py:256]: Tiny quibble: I personally prefer positive assertions for types so it's easier to understand what `n.target` is supposed to be. Also, from having to debug these assertions going sour, I tend to put something to print in the second parameter, in the most extreme case it would go like this:

`assert isinstance(n.target, (foo, bar, baz)), (n.target, type(n.target))`

but most often

`assert isinstance(n.target, (foo, bar, baz)), n.target`

**Review by rec — APPROVED** [review]
Well, looking back at these comments, they're all quibble of some variety or other, so if you ignored them all, I wouldn't care.

This seems like a solid step forward for typing which will have positive ramifications all over the code. I'm a little surprised there weren't even more changes in distant files!

Good stuff.

**Comment by rec** [comment]
Oh, and tangentially relevant: in another code review I have this `TypeAlias` in `_inductor/ir.py`.

```
_OpOverloads: TypeAlias = Union[torch._ops.OpOverload, torch._ops.HigherOrderOperator]
```

It passed review but I'm not entirely happy with the name - do you have a better one?

File: torch/_ops.py (hunk 3 above)
Comment (amjames) [reply torch/_ops.py:1356]: Sounds reasonable to me.

File: torch/ao/quantization/fx/_lower_to_native_backend.py [hunk 11]
```diff
@@ -28,7 +28,7 @@
 )
 
 
-QOP_TO_ARG_NAMES_TO_SKIP = {
+QOP_TO_ARG_NAMES_TO_SKIP: dict[Callable, list[str]] = {
```
Comment (amjames) [inline torch/ao/quantization/fx/_lower_to_native_backend.py:31]: Would it not be better to do `dict[OpOverloadPacket, list[str]]`?

File: torch/_ops.py (hunk 2 above)
Comment (benjaminglass1) [reply torch/_ops.py]: Done!

File: torch/_ops.py (hunk 6 above)
Comment (benjaminglass1) [reply torch/_ops.py:751]: @rec Weirdly, `mypy` seems to flag some instances of `Callable`, but not others. I have not figured out any pattern to where it's annoyed. I'll tweak these.

File: torch/fx/node.py (hunk 9 above)
Comment (benjaminglass1) [reply torch/fx/node.py:89]: @rec I'm truly confused here; the best I can tell is that `mypy` struggles with auto-detecting `Callable` from mixed sets with different classes of callables, and defaults to `object`. I tried a number of different ideas, and eventually gave up and ignored this one specific place.

Comment (rec) [reply torch/fx/node.py:89]: I gotta try this, this is infuriating. :-D BRB!

File: torch/_ops.py [hunk 12]
```diff
@@ -1288,19 +1306,18 @@ class _OpNamespace(types.ModuleType):
         operation will already exist).
     """
 
-    def __init__(self, name):
+    __file__ = "torch.ops"
```
Comment (Skylion007) [inline torch/_ops.py:1309]: typing here too

File: torch/_ops.py [hunk 13]
```diff
@@ -1353,17 +1370,22 @@ def _refresh_packet(packet):
     packet._overload_names = overload_names
 
 
-class _PyOpNamespace(_OpNamespace):
-    def __init__(self, name, ops):
-        super().__init__(name)
-        self._ops = ops
+class _HigherOrderNamespace(types.ModuleType):
+    __file__ = "torch.ops"
```
Comment (Skylion007) [inline torch/_ops.py:1374]: and here, etc

**Review by Skylion007 — APPROVED** [review]
Aside from the naked callables, this looks great.

File: torch/_ops.py (hunk 12 above)
Comment (benjaminglass1) [reply torch/_ops.py:1309]: @Skylion007 This is the one that `mypy` registered the base class having `Literal` type on. I'd argue that we shouldn't be adding that type to our code. If you would still prefer it, I will though.

File: torch/fx/node.py (hunk 9 above)
Comment (benjaminglass1) [reply torch/fx/node.py:89]: @rec suggested a way to improve this somewhat, but this represents a bug in `mypy`. I've opened an issue at https://github.com/python/mypy/issues/19091 and will resolve this after committing @rec's suggestion.

Comment (Skylion007) [reply torch/fx/node.py:89]: Does using a 
`_F = TypeVar("_F", bound="Callable[Any, ...]")`
solve the issue?

File: torch/_ops.py (hunk 12 above)
Comment (Skylion007) [reply torch/_ops.py:1309]: Oh because, it's a module type, yeah it makes sense it's a literal then... This mypy telling you should probably shouldn't override this lol

File: torch/fx/node.py (hunk 9 above)
Comment (benjaminglass1) [reply torch/fx/node.py:89]: @Skylion007 No, that would be an unbound `TypeVar` in this context. What I've switched to is simply `*_side_effectful_need_to_be_preserved_pre_dispatch`, which apparently works.

File: torch/fx/passes/reinplace.py (hunk 10 above)
Comment (benjaminglass1) [reply torch/fx/passes/reinplace.py:256]: @rec This one is actually necessarily a negative assertion, because the other possibility is `Callable[..., Any]`, which can't be directly asserted for as a generic type.

**Comment by benjaminglass1** [comment]
> Oh, and tangentially relevant: in another code review I have this `TypeAlias` in `_inductor/ir.py`.
> 
> ```
> _OpOverloads: TypeAlias = Union[torch._ops.OpOverload, torch._ops.HigherOrderOperator]
> ```
> 
> It passed review but I'm not entirely happy with the name - do you have a better one, given that you're thinking about these sorts of things?

@rec You could simply use `torch._ops.OperatorBase`, I believe?

File: torch/ao/quantization/fx/_lower_to_native_backend.py (hunk 11 above)
Comment (benjaminglass1) [reply torch/ao/quantization/fx/_lower_to_native_backend.py:31]: That would be great, but the place where we actually query the dictionary can only assert that the key is `Callable[..., Any]` at best, so this typing needs to be more generic to avoid triggering warnings about that.

File: torch/_ops.py [hunk 14]
```diff
@@ -735,7 +745,14 @@ def get_cached_ops():
 # Each OpOverload object contains pointer to a specific operator overload, a pointer to the parent `OpOverloadPacket` object.
 # You can obtain an OpOverload object through attribute query on OpOverloadPacket.
 class OpOverload(OperatorBase):
-    def __init__(self, overloadpacket, op, op_dk, schema, tags):
+    def __init__(
+        self,
+        overloadpacket: "OpOverloadPacket",
+        op: Callable[..., Any],
```
Comment (Skylion007) [inline torch/_ops.py:751]: still naked callables here, not great.

File: torch/_ops.py [hunk 15]
```diff
@@ -1098,14 +1112,22 @@ def _has_script_object_arg(schema: torch.FunctionSchema) -> bool:
...
 # You can obtain an OpOverload object through attribute query.
 class OpOverloadPacket:
-    def __init__(self, qualified_op_name, op_name, op, overload_names):
+    __file__: ClassVar[str] = "torch.ops"
+
+    def __init__(
+        self,
+        qualified_op_name: str,
+        op_name: str,
+        op: Callable[..., Any],
```
Comment (Skylion007) [inline torch/_ops.py:1121]: same

File: torch/_ops.py (hunk 14 above)
Comment (benjaminglass1) [reply torch/_ops.py:751]: @Skylion007 I would prefer to switch it to something more concrete, but after a good bit of experimenting I was unable to find anything other than `Callable[..., Any]` that worked in all the ways this was invoked. The next best solution was `Callable[_P, _T]`, with `op_dk` getting `Callable[_Q, _U]`, which didn't seem particularly informative to anyone, since it specified no similarities between the two callables.

Comment (benjaminglass1) [reply torch/_ops.py:751]: Actually, I think I fundamentally misunderstood something I was trying to use to inform my typing of this. Hang on a minute, I may have a type for this.

Comment (benjaminglass1) [reply torch/_ops.py:751]: @Skylion007 I'm about to push something that explicitly links the types of `op` and `op_dk` here in this initializer only. I tried locally to make things even more generic, but I think we should stop here:

1. Genericizing `op` in the initializer for `OpOverloadPacket` doesn't add any additional clarity, since the generics aren't linked to anything that couldn't be inferred otherwise.
2. Fully genericizing the classes `OpOverload` and `OpOverloadPacket` solves problem 1, but you have to apply generics everywhere in the codebase that references either class in a type. That's a lot of boilerplate code for something that will almost always be filled in with generic types (since very rarely do we a priori know which arg types and return types to expect in functions being typed with these classes).

File: torch/_ops.py (hunk 15 above)
Comment (benjaminglass1) [reply torch/_ops.py:1121]: See above.

**Review by cyyever — APPROVED** [review]

File: torch/_ops.py (hunk 14 above)
Comment (benjaminglass1) [reply torch/_ops.py]: @Skylion007 I'm going to try to merge this; if this is still an issue, ping me, and I'll address in a follow-on PR.

Comment (Skylion007) [reply torch/_ops.py]: Doesn't this lead to type erasure of the arguments of the callabke without ParamSpec (or without treating op as it's own typevar)?

Comment (benjaminglass1) [reply torch/_ops.py]: That would be a concern, but all the callables that are getting passed to this constructor (unless I've completely misunderstood something :sweat_smile:) are runtime-loaded compiled ops, so we never had types to begin with.

**Comment by yangw-dev** [comment]
@benjaminglass1 FYI we revert due to internal test failures

**Comment by Skylion007** [comment]
@aorenste Would you be able to help shepherd this change in too?

**Comment by aorenste** [comment]
Superceded by #154555

**Comment by aorenste** [comment]
> @aorenste I'll close this one! Would you mind listing me as a co-author on #154555?

I certainly don't mind - you did all the work. But do you have any idea how to do that?
